src/utils/LookmlDiagrammer/diagrammer.ts

```diff
@@ -0,0 +1,259 @@
+      return datum
+    })
+    ]
+  })
+  
+  // Add join data to DiagramDict for each join
+  diagramDict.joinData = exploreJoins.map((join: ILookmlModelExploreJoins, joinIndex: number) => {
+    let joinPath: DiagramJoin[] = []
+    if (join.foreign_key) {
+      let baseTableId = join.foreign_key.includes(".") 
+        ? join.foreign_key.split(".")[0] 
+        : explore.name
+      let baseTableRef = diagramDict.tableData[baseTableId]
+      let baseTableLookupValue = join.foreign_key.includes(".")
+        ? join.foreign_key 
+        : (explore.name + "." + join.foreign_key)
+      let fieldIndex = getViewFieldIndex(baseTableRef, baseTableLookupValue)
+      let pkTableRef = diagramDict.tableData[join.name]
+      let pkFieldIndex = getViewPkIndex(pkTableRef)
+
+      joinPath.push(getPkJoinPathObj(join, pkTableRef, pkFieldIndex))
+      joinPath.push(getBaseJoinPathObj(join, baseTableId, baseTableRef, fieldIndex))
+    }
+    else if (join.dependent_fields.length > 0) {
+      join.dependent_fields.forEach((field: string, depFieldIndex: number) => {
+        let joinFieldArr = field.split(".")
+        let tableRef = diagramDict.tableData[joinFieldArr[0]]
+        let fieldIndex = getViewDependentFieldIndex(tableRef, field)
+        if (join.sql_on) {
+          join.sql_on.includes("${"+field+"}") 
+            && joinPath.push(getSqlJoinPathObj(joinFieldArr, fieldIndex, field, join))
+        } else {
+          joinPath.push(getSqlJoinPathObj(joinFieldArr, fieldIndex, field, join))
+          joinPath.push(getJoinPathObj(join))
+        }
+      })
+    } else {
+      joinPath.push(getJoinPathObj(join))
+      joinPath.push(getExploreJoinPathObj(explore, join))
+    }
+    return joinPath
+  })
+
+  // General order tables would be arranged in, if no joins and no base view
+  const joinCount = countJoins(diagramDict)
+  let buildOrder = views.sort((a: string, b: string)=>{
+    return joinCount[a] > joinCount[b] ? -1 : 1;
+  })
+
+  // For each table, get list of tables joined by way of it
+  const scaffold = getJoinedViewsForView(buildOrder, diagramDict, explore)
+
+  // list of view names that have been arranged
+  let built: string[] = []
+
+  // keeps track of the vertical distance which has
+  // been filled by previous tables for each degree
+  let shift: any = {}
+  
+  // keeps track of the order the tables
+  // are arranged in vertically for each degree
+  let yOrder: any = {}
+
+  /**
+   * A recursive function for arranging diagram tables. Starting at the base table,
+   * arrange each of its each joined tables to the left and right, switching sides for each
+   * table and leaving a gap between tables of same degree. As a table is arranged, join any of its 
+   * tables in the direction it was placed off of the base table. 
+   * @param table - view name
+   * @param degree - number of joins away from base
+   */
+  function arrangeTables(table: string, degree: number) {
```

Comment:
Having difficulty figuring out how to extract this function bc it is so entangled. 

It's also the heart of the application

Perhaps it should stay? Please advise

Reply:
Oh boy, thats a doozy of a function :D.

1. Consider converting arrangeTables to a class (TableArranger?). scaffold, built, shift and yOrder become properties of the class.
2. Consider passing in scaffold, built, shift and yOrder as arguments to the function arrangeTables and pass them down recursively.

Reply:
Wow, I really like the class based implementation. TBH it's my first in JS :smiley_cat: 

Over time I will move more functionality into this class, thus more generic naming.

Reply:
Before you add more functionality to the class consider the single responsibility principle. You may be better off have more classes. Remember that you can have compositions of classes (as an aside, composition better than inheritance, at least in my opinion).

Reply:
Yes, I would want LookmlDiagrammer as orchestration class that invokes JoinUntangler TableArranger ViewBuilder etc

Reply:
Future work for now tho. More diagrammer classes would help in implementing Edit mode (v2)